File: ts/src/lib/parsers/movie.ts
import { ReadableStream } from "@open-flash/stream";
import { Incident } from "incident";
import { inflate } from "pako";
import { Uint8 } from "semantic-types";
import { CompressionMethod, Header, Movie, SwfSignature, Tag } from "swf-types";
import { parseHeader, parseSwfSignature } from "./header";
import { parseTagBlockString } from "./tags";

/**
 * Parses a completely loaded SWF file.
 *
 * @param byteStream SWF stream to parse
 */
export function parseSwf(byteStream: ReadableStream): Movie {
  const signature: SwfSignature = parseSwfSignature(byteStream);
  switch (signature.compressionMethod) {
    case CompressionMethod.None:
      return parseMovie(byteStream, signature.swfVersion);
    case CompressionMethod.Deflate:
      const tail: Uint8Array = byteStream.tailBytes();
      const payload: Uint8Array = inflate(tail);
      const payloadStream: ReadableStream = new ReadableStream(payload);
      return parseMovie(payloadStream, signature.swfVersion);
    case CompressionMethod.Lzma:
      throw new Incident("NotImplemented", "Support for LZMA compression is not implemented yet");
    default:
      throw new Incident("UnknownCompressionMethod", "Unknown compression method");
  }
}

/**
 * Parses a completely loaded movie.
 *
 * The movie is the uncompressed payload of the SWF.
 *
 * @param byteStream Movie bytestream
 * @param swfVersion Parsed movie.
 */
export function parseMovie(byteStream: ReadableStream, swfVersion: Uint8): Movie {
  const header: Header = parseHeader(byteStream, swfVersion);
  const tags: Tag[] = parseTagBlockString(byteStream, swfVersion);
  return {header, tags};
}

File: rs/src/parsers/movie.rs
use crate::parsers::header::{parse_header, parse_swf_signature};
use crate::parsers::tags::parse_tag;
use crate::state::ParseState;
use nom::{IResult as NomResult, Needed};
use swf_tree as ast;

pub fn parse_tag_block_string<'a>(input: &'a [u8], state: &ParseState) -> NomResult<&'a [u8], Vec<ast::Tag>> {
  let mut result: Vec<ast::Tag> = Vec::new();
  let mut current_input: &[u8] = input;
  while current_input.len() > 0 {
    // TODO: Check two bytes ahead
    // A null byte indicates the end of the string of tags
    if current_input[0] == 0 {
      current_input = &current_input[1..];
      break;
    }
    match parse_tag(current_input, state) {
      Ok((next_input, swf_tag)) => {
        current_input = next_input;
        result.push(swf_tag);
      }
      Err(::nom::Err::Incomplete(_)) => return Err(::nom::Err::Incomplete(Needed::Unknown)),
      Err(e) => return Err(e),
    };
  }
  Ok((current_input, result))
}

pub fn parse_movie_payload(input: &[u8], swf_version: u8) -> NomResult<&[u8], ast::Movie> {
  let mut state = ParseState::new(swf_version);
  let (input, header) = parse_header(input, swf_version)?;
  let (input, tags) = parse_tag_block_string(input, &mut state)?;

  Ok((input, ast::Movie { header, tags }))
}

pub fn parse_movie(input: &[u8]) -> NomResult<&[u8], ast::Movie> {
  use ::std::io::Write;

  let (input, signature) = parse_swf_signature(input)?;
  match signature.compression_method {
    ast::CompressionMethod::None => parse_movie_payload(input, signature.swf_version),
    ast::CompressionMethod::Deflate => {
      let mut decoder = ::inflate::InflateWriter::from_zlib(Vec::new());
      decoder.write(input).unwrap();
      let payload = decoder.finish().unwrap();

      match parse_movie_payload(&payload[..], signature.swf_version) {
        Ok((_, movie)) => Ok((&[][..], movie)),
        Err(::nom::Err::Error((_, e))) => Err(::nom::Err::Error((&[][..], e))),
        Err(::nom::Err::Failure((_, e))) => Err(::nom::Err::Failure((&[][..], e))),
        Err(::nom::Err::Incomplete(n)) => Err(::nom::Err::Incomplete(n)),
      }
    }
    ast::CompressionMethod::Lzma => unimplemented!(),
  }
}
